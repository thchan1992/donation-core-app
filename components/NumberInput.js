import { StyleSheet, Text, TextInput, View } from "react-native";

const NumberInput = ({
  value,
  setter,
  placeholder,
  height,
  maxWidth,
  minWidth,
  maxLength,
  isPound,
}) => {
  return (
    <View style={{ flexDirection: "row" }}>
      {isPound != null && <Text style={styles.poundsSign}>Â£</Text>}
      <TextInput
        textAlign={"center"}
        placeholderTextColor={"grey"}
        onChangeText={setter}
        maxLength={maxLength}
        style={[
          styles.input,
          {
            minheight: height,
            fontSize: height,
            maxWidth: maxWidth,
            minWidth: minWidth,
          },
        ]}
        value={value}
        placeholder={placeholder}
        keyboardType="numeric"
      />
    </View>
  );
};

export default NumberInput;

const styles = StyleSheet.create({
  input: {
    margin: 12,
    borderWidth: 1,
    padding: 10,
    borderColor: "#FF304F",
    borderRadius: 8,
    minheight: 50,
    fontSize: 50,
    color: "white",
    fontSize: 50,
    maxWidth: 200,
  },
  poundsSign: {
    fontSize: 50,
    color: "white",
    paddingTop: 20,
  },
});
