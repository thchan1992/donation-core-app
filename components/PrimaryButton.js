import { TouchableOpacity, Text, StyleSheet } from "react-native";
// PrimaryButton is a functional component that renders a button with a text and onPress functionality
const PrimaryButton = ({ text, onPress }) => {
  return (
    // TouchableOpacity is used to make the button responsive to touch/press events
    <TouchableOpacity style={styles.button} onPress={onPress}>
      <Text style={styles.text}>{text}</Text>
    </TouchableOpacity>
  );
};

export default PrimaryButton;

const styles = StyleSheet.create({
  button: {
    width: 150,
    height: 70,
    margin: 10,
    borderRadius: 8,
    backgroundColor: "#28C7FA",
    borderRadius: 8,
    borderWidth: 1.5,
    borderColor: "#FF304F",
  },
  text: {
    alignSelf: "center",
    padding: 16,
    fontWeight: "bold",
    fontSize: 30,
  },
});
