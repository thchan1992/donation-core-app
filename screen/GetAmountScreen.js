import { useState } from "react";
import {
  View,
  Text,
  StyleSheet,
  Image,
  TextInput,
  ScrollView,
} from "react-native";
import PrimaryButton from "../components/PrimaryButton";
import NumberInput from "../components/NumberInput";

const GetAmountScreen = ({ navigation, route }) => {
  const [donateAmount, setDonateAmount] = useState(null);
  const { charity } = route.params;

  const handleSubmit = () => {
    navigation.navigate("CardPaymentScreen", {
      charity: charity,
      donateAmount: donateAmount,
    });
  };

  return (
    <View style={styles.container}>
      <Image style={styles.image} source={charity.imageUrl} />
      <ScrollView>
        <Text style={styles.instrucText}>
          How much would you like to donate to {charity.name}?
        </Text>
        <View style={styles.inputContainer}>
          <NumberInput
            value={donateAmount}
            setter={setDonateAmount}
            placeholder={"Amount"}
            height={40}
            maxWidth={160}
            minWidth={160}
            maxLength={5}
            isPound={true}
          />
          <PrimaryButton text={"Donate"} onPress={handleSubmit} />
        </View>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    paddingTop: 10,
  },
  input: {
    height: 40,
    margin: 12,
    borderWidth: 1,
    padding: 10,
  },
  inputContainer: {
    alignItems: "center",
    justifyContent: "center",
  },
  image: {
    width: 250,
    height: 250,
    borderWidth: 3,
    borderColor: "#FF304F",
    borderRadius: 8,
    padding: 30,
  },
  instrucText: {
    alignSelf: "center",
    padding: 5,
    fontWeight: "bold",
    fontSize: 25,
    color: "white",
  },
});

export default GetAmountScreen;
