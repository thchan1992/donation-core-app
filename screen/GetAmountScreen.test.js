import React from "react";
import renderer from "react-test-renderer";
import { Alert } from "react-native";
import GetAmountScreen, { styles } from "./GetAmountScreen";
import PrimaryButton from "../components/PrimaryButton";
import NumberInput from "../components/NumberInput";

jest.mock("react-native/Libraries/Animated/NativeAnimatedHelper");
jest.mock("react-native/Libraries/Alert/Alert", () => {
  return {
    alert: jest.fn(),
  };
});

describe("GetAmountScreen", () => {
  const mockNavigation = { navigate: jest.fn() };
  const mockRoute = {
    params: {
      charity: {
        name: "Test Charity",
        imageUrl: require("../assets/adaptive-icon.png"),
        info: "Test charity information",
      },
    },
  };

  it("renders correctly with default state", () => {
    const tree = renderer
      .create(<GetAmountScreen navigation={mockNavigation} route={mockRoute} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('navigates to "CardPaymentScreen" with correct params when "Donate" button is pressed and donation amount is valid', () => {
    const component = renderer.create(
      <GetAmountScreen navigation={mockNavigation} route={mockRoute} />
    );
    const instance = component.root;
    const numberInput = instance.findByType(NumberInput);
    numberInput.props.setter(5);
    const donateButton = instance.findByType(PrimaryButton);
    donateButton.props.onPress();
    expect(mockNavigation.navigate).toHaveBeenCalledWith("CardPaymentScreen", {
      charity: mockRoute.params.charity,
      donateAmount: 5,
    });
  });

  it('shows alert when "Donate" button is pressed and donation amount is invalid', () => {
    const component = renderer.create(
      <GetAmountScreen navigation={mockNavigation} route={mockRoute} />
    );
    const instance = component.root;
    const numberInput = instance.findByType(NumberInput);
    numberInput.props.setter(null);
    const donateButton = instance.findByType(PrimaryButton);
    donateButton.props.onPress();
    expect(Alert.alert).toHaveBeenCalled();
  });
});
