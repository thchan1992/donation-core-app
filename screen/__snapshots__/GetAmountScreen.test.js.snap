// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetAmountScreen renders correctly with default state 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "paddingTop": 10,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "borderColor": "#FF304F",
            "borderRadius": 8,
            "borderWidth": 3,
            "height": 250,
            "padding": 30,
            "width": 250,
          }
        }
      />
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "white",
            "fontSize": 25,
            "fontWeight": "bold",
            "padding": 5,
          }
        }
      >
        Please insert the amount you would like to donate to Test Charity. The amount should not be larger than £1000
      </Text>
      <View
        style={
          Array [
            Object {
              "height": 400,
            },
            Object {
              "alignItems": "center",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 50,
                "paddingTop": 20,
              }
            }
          >
            £
          </Text>
          <TextInput
            keyboardType="numeric"
            maxLength={6}
            onChangeText={[Function]}
            placeholder="Amount"
            placeholderTextColor="grey"
            style={
              Array [
                Object {
                  "borderColor": "#FF304F",
                  "borderRadius": 8,
                  "borderWidth": 1,
                  "color": "white",
                  "fontSize": 50,
                  "margin": 12,
                  "maxWidth": 200,
                  "minheight": 50,
                  "padding": 10,
                },
                Object {
                  "fontSize": 40,
                  "maxWidth": 160,
                  "minWidth": 160,
                  "minheight": 40,
                },
              ]
            }
            textAlign="center"
            value={null}
          />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#28C7FA",
              "borderColor": "#FF304F",
              "borderRadius": 8,
              "borderWidth": 1.5,
              "height": 70,
              "margin": 10,
              "opacity": 1,
              "width": 150,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "fontSize": 30,
                "fontWeight": "bold",
                "padding": 16,
              }
            }
          >
            Donate
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
